src/www/Inset.ts

```diff
@@ -0,0 +1,266 @@
+     * 
+     * This is the equivilant of calling Inset.free(insetInstance)
+     * 
+     * @returns 
+     */
+    public async free(): Promise<void> {
+        return await Inset.free(this);
+    }
+
+    /**
+     * Adds a listener to this inset configuration.
+     * 
+     * Note that this may fire even if nothing has actually
+     * changed.
+     * 
+     * Retain the listener reference to remove it later if
+     * necessary.
+     * 
+     * @param listener 
+     */
+    public addListener(listener: IInsetCallbackFunc): void {
+        this.$listeners.push(listener);
+        listener(this.$currentInsets);
+    }
+
+    /**
+     * Frees the listener reference
+     * 
+     * @param listener 
+     */
+    public removeListener(listener: IInsetCallbackFunc): void {
+        let idx: number = this.$listeners.indexOf(listener);
+        if (idx > -1) {
+            this.$listeners.splice(idx, 1);
+        }
+    }
+
+    public onUpdate(insets: IInset): void {
+        this.$currentInsets = insets;
+
+        for (let i = 0; i < this.$listeners.length; i++) {
+            this.$listeners[i](insets);
+        }
+    }
+
+    private static $generateID(): string {
+        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
+            const r = Math.random() * 16 | 0,
+                v = c === 'x' ? r : (r & 0x3 | 0x8);
+            return v.toString(16);
+        });
+    }
+
+    /**
+     * Configures a new Inset instance to listen for inset changes
+     * It's valid to have multiple instances, with different configurations
+     * Each instance may have 0-to-many listeners attached via addListener
+     * 
+     * If this instance is no longer needed/used, call `free` to free
+     * resources.
+     * 
+     * It will be more performant to keep the instance count low. If only one
+     * configuration set is needed, then it would be recommended to create a
+     * single instance and share it rather than every object having it's own
+     * inset listener instance.
+     * 
+     * @param config 
+     * @returns 
+     */
+    public static create(config: IInsetConfiguration): Promise<Inset> {
+        return new Promise<Inset>((resolve, reject) => {
+            if (!config) {
+                config = {};
+            }
+
+            if (cordova.platformId === 'ios') {
+                let instance: Inset = new Inset(Inset.$generateID());
+                Inset.$instances.set(instance.getID(), instance);
+                resolve(instance);
+                return;
+            }
+
+            cordova.exec(
+                (e: IInsetEvent) => {
+                    if (Inset.$isInitEvent(e)) {
+                        let instance: Inset = new Inset(e.data); 
+                        Inset.$instances.set(instance.getID(), instance);
+                        resolve(instance);
+                    }
+                    else if (Inset.$isUpdateEvent(e)) {
```

Comment:
These methods are typescript with a special signature so that it can change the type of `e` if it returns true.

That's why the method has a `e is X` return type. That way TypeScript will treat `e` as `X` while inside the if condition block.

It cannot be replaced with a constant or enum.